flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/NettyMessageClientDecoderDelegateTest.java

```diff
@@ -0,0 +1,411 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.netty;
+
+import org.apache.flink.core.memory.MemorySegment;
+import org.apache.flink.core.memory.MemorySegmentFactory;
+import org.apache.flink.runtime.io.network.PartitionRequestClient;
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.io.network.buffer.FreeingBufferRecycler;
+import org.apache.flink.runtime.io.network.buffer.NetworkBuffer;
+import org.apache.flink.runtime.io.network.buffer.NetworkBufferPool;
+import org.apache.flink.runtime.io.network.partition.consumer.InputChannelID;
+import org.apache.flink.runtime.io.network.partition.consumer.RemoteInputChannel;
+import org.apache.flink.runtime.io.network.partition.consumer.SingleInputGate;
+import org.apache.flink.shaded.netty4.io.netty.buffer.ByteBuf;
+import org.apache.flink.shaded.netty4.io.netty.channel.embedded.EmbeddedChannel;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static junit.framework.TestCase.assertEquals;
+import static junit.framework.TestCase.assertTrue;
+import static org.apache.flink.runtime.io.network.netty.NettyMessage.BufferResponse;
+import static org.apache.flink.runtime.io.network.netty.NettyMessage.ErrorResponse;
+import static org.apache.flink.runtime.io.network.partition.InputChannelTestUtils.createRemoteInputChannel;
+import static org.apache.flink.runtime.io.network.partition.InputChannelTestUtils.createSingleInputGate;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.fail;
+import static org.mockito.Mockito.mock;
+
+/**
+ * Tests the client side message decoder.
+ */
+public class NettyMessageClientDecoderDelegateTest {
+	private static final NettyBufferPool ALLOCATOR = new NettyBufferPool(1);
+
+	/**
+	 * Verifies that the client side decoder works well for unreleased input channels.
+	 */
+	@Test
+	public void testDownstreamMessageDecode() throws Exception {
+		// 6 buffers required for running 2 rounds and 3 buffers each round.
+		NettyChannelAndInputChannelIds context = createPartitionRequestClientHandler(6);
+
+		Supplier<NettyMessage[]> messagesSupplier = () -> {
+			Buffer event = createDataBuffer(32);
+			event.tagAsEvent();
+
+			return new NettyMessage[]{
+				new NettyMessage.BufferResponse(createDataBuffer(128), 0, context.getNormalChannelId(), 4),
+				new NettyMessage.BufferResponse(createDataBuffer(256), 1, context.getNormalChannelId(), 3),
+				new NettyMessage.BufferResponse(event, 2, context.getNormalChannelId(), 4),
+				new NettyMessage.ErrorResponse(new RuntimeException("test"), context.getNormalChannelId()),
+				new NettyMessage.BufferResponse(createDataBuffer(56), 3, context.getNormalChannelId(), 4)
+			};
+		};
+
+		repartitionMessagesAndVerifyDecoding(
+			context,
+			messagesSupplier,
+			(int[] sizes) -> new int[]{
```

Comment:
It is hard to understand the rules to split the message and it is done differently in different tests.
I suggest simplifying this logic to split the message in fixed-length, E.g. the size of `BufferResponse` is 128, then we can split it based on fixed 30. Then we do not need to pass this argument which can be hidden inside the split logic.

Reply:
Changed to logic to use a fixed partition method.